datafusion/common/src/scalar/mod.rs

```diff
@@ -1714,6 +1631,15 @@ impl ScalarValue {
             Some(sv) => sv.data_type(),
         };
 
+        Self::iter_to_array_of_type(scalars.collect(), &data_type)
+    }
+
+    fn iter_to_array_of_type(
```

Comment:
This was introduced because of aggregations that use ScalarValue as state and so they could lose type information when emitting state.